vendor/bat-native-ledger/src/bat/ledger/internal/promotion/promotion.cc

```diff
@@ -251,14 +243,34 @@ void Promotion::OnGetAllPromotions(
       continue;
     }
 
-    promotions.insert(std::make_pair(item->id, item->Clone()));
+    promotions_ui.push_back(item->Clone());
 
     ledger_->database()->SavePromotion(
         item->Clone(),
         [](const ledger::Result _){});
   }
 
-  ledger::PromotionList promotions_ui = GetPromotionsForUI(promotions);
+  bool found;
+  for (const auto& promotion : promotions) {
+    if (promotion.second->status != ledger::PromotionStatus::ACTIVE) {
+      break;
+    }
+
+    found = false;
+    for (const auto& item : list) {
```

Comment:
Another option would be something like:

```cpp
bool found = std::any_of(list.begin(), list.end(), [&promotion](auto& item) {
  item->id == promotion.second->id;
});
```

But haven't we already erased all entries in `promotions` that have a matching id (on line 255)? Is this search required?

Reply:
yes it's needed. We only erased promotions that we received from api and we have it in db. What we are doing here is going through active promotions that we have in db and if we don't find them in api we mark them as over

Reply:
Thinking this through...

The loop starting on line 21 goes through all entries in `list` and removes the corresponding entry in `promotions`, if there is one. So after the loop completes, all entries in `list` have been removed from `promotions`, and `promotions` only contains entries that do not have a matching entry in `list`. Am I missing something?

Another way to simplify things (while avoiding a nested loop) might be to *not* erase from `promotions`, and instead populate an `std::set` with the ids from `list`, and then check that set in this loop over `promotions`.

Reply:
that is correct. Use case is like this. Server serves promotion and we put it in db. Promotion has status active. User doesn't claim it. Then this promotion is removed from the server. We need to remove all promotions from db that has status active, but don't have corresponding promotion from server response